common/persistence/nosql/nosqlplugin/interfaces.go

```diff
@@ -203,6 +204,119 @@ type (
+	 */
+	taskCRUD interface {
+		// SelectTaskList returns a single tasklist row.
+		// Return IsNotFoundError if the row doesn't exist
+		SelectTaskList(ctx context.Context, filter *TaskListFilter) (*TaskListRow, error)
+		// InsertTaskList insert a single tasklist row
+		// Return IsConditionFailedError if the row already exists, and also the existing row
+		InsertTaskList(ctx context.Context, row *TaskListRow) (previous *TaskListRow, err error)
+		// UpdateTaskList updates a single tasklist row
+		// Return IsConditionFailedError if the condition doesn't meet, and also the previous row
+		UpdateTaskList(ctx context.Context, row *TaskListRow, previousRangeID int64) (previous *TaskListRow, err error)
+		// UpdateTaskList updates a single tasklist row, and set an TTL on the record
+		// Return IsConditionFailedError if the condition doesn't meet, and also the existing row
+		// Ignore TTL if it's not supported, which becomes exactly the same as UpdateTaskList, but ListTaskList must be
+		// implemented for TaskListScavenger
+		UpdateTaskListWithTTL(ctx context.Context, ttlSeconds int64, row *TaskListRow, previousRangeID int64) (previous *TaskListRow, err error)
+		// ListTaskList returns all tasklists.
+		// Noop if TTL is already implemented in other methods
+		ListTaskList(ctx context.Context, pageSize int, nextPageToken []byte) (*ListTaskListResult, error)
+		// DeleteTaskList deletes a single tasklist row
+		// Return IsConditionFailedError if the condition doesn't meet, and also the existing row
+		DeleteTaskList(ctx context.Context, filter *TaskListFilter, previousRangeID int64) (*TaskListRow, error)
+		// InsertTasks inserts a batch of tasks
+		// Return IsConditionFailedError if the condition doesn't meet, and also the previous tasklist row
+		InsertTasks(ctx context.Context, tasksToInsert []*TaskRowForInsert, tasklistCondition *TaskListRow) (previous *TaskListRow, err error)
+		// SelectTasks return tasks that associated to a tasklist
+		SelectTasks(ctx context.Context, filter *TasksFilter) ([]*TaskRow, error)
+		// DeleteTask delete a single task
+		DeleteTask(ctx context.Context, row *TaskRowPK) error
+		// DeleteTask delete a batch tasks that taskIDs less than the row(exclusive)
+		// Limit is the max number of tasks to delete.
+		// Also return the number of rows deleted -- if it's not supported then ignore the limit, and return persistence.UnknownNumRowsAffected
+		RangeDeleteTasks(ctx context.Context, maxTaskID *TaskRowPK, limit int) (rowsDeleted int, err error)
+	}
+
+	TasksFilter struct {
+		TaskListFilter
+		// Inclusive
+		MinTaskID int64
+		// Exclusive
+		MaxTaskID int64
+		BatchSize int
+	}
+
+	TaskRowForInsert struct {
+		TaskRow
+		// <= 0 means no TTL
+		TTLSeconds int
+	}
+
+	TaskRow struct {
+		DomainID     string
+		TaskListName string
+		TaskListType int
+		TaskID       int64
+
+		WorkflowID  string
+		RunID       string
+		ScheduledID int64
+		CreatedTime time.Time
+	}
+
+	TaskRowPK struct {
```

Comment:
Can we reuse `TasksFilter`? Also not sure if `BatchSize` belongs to `TasksFilter` logically.

Reply:
Yeah that's a bit tricky here. TasksFilter has `MinTaskID` which is not supported in this API. Other than this, we have several options:
1. Reuse it any way, and tell the developer to ignore `MinTaskID` in this API. 
2. Reuse it, and implement it, but we don't have tests to cover this, as the persistenceIntefaces+dataInterfaces doesn't have `MinTaskID` in rangeDelete. Unless we also change the persistenceIntefaces and dataInterfaces. We probably don't want to do this in this refactoring. 
3. Create a different taskFilter, e.g. `TasksFilterNoMin`. 

I chose this TaskRowPK because 1. it can be resued in other methods 2. It help indicating what is PrimaryKey in this table. 
Do you have any preference?

Reply:
I think the most elegant way is to use TasksFilter like you said. So I changed to option 2, and create an issue https://github.com/uber/cadence/issues/4243

Reply:
I am not sure if there's any concern from performance side when using rangeDelete to delete a single task. But we always handle the special case of deleting only one task with a point delete query in the plugin implementation.

Reply:
It depends on the DB implementation, but in theory this should not be a concern. 
Also, this CompleteTask data API is only being used by orphan task scanner, which should be very rarely used.